### ISSUE TITLE
React flow rendering custom nodes outside reactflowcontext

### ISSUE THREAD

**Issue by janarvaezkng** [xyflow/xyflow#4884]
### What platform were you using when you found the bug?

- React Flow version: 12.3.6
- Browser and version: Chrome latest
- OS and version: macos latest

### Describe the Bug

I have a library that creates a Tree flow using React flow.

```ts
export const MyTree = forwardRef<MyTreeRef, MyTreeProps>(
  function MyTree(props, ref) {
    return (
      <ReactFlowProvider>
        <MyTreeUnwrapped ref={ref} {...props} />
      </ReactFlowProvider>
    );
  }
);

const MyTreeUnwrapped = forwardRef<MyTreeRef, MyTreeProps>(
   function MyTreeUnwrapped(props, ref) {

   return (
      <MyTreeContext.Provider
        value={...}
      >
          <ReactFlow
            nodes={nodes}
            edges={edges}
            onNodesChange={onNodesChange}
            onEdgesChange={onEdgesChange}
          >
            ...
          </ReactFlow>
      </MyTreeContext.Provider>
    );
});
```

When creating custom nodes within the library, everything works as expected.
But when using this library within another App, if I create a custom node in the App itself and pass it as a prop (then same way I do it within the library), then I get this error:

> React Flow]: Seems like you have not used zustand provider as an ancestor. Help: https://reactflow.dev/error#001

The only solution if to wrap my tree component inside another reactflowprovider like:

```ts
<ReactFlowProvider>
    <MyTree customNodes={customNodes} ... />
</ReactFlowProvider>
```

I traced the error to the use of `<Handle />` components in my custom nodes, most probably if I were using reactflow hooks in the nodes I would get the error as well.

For some reason it looks like it does a first render outside the context, and then it works fine, since I get a lot of warnings about:
> [React Flow]: Handle: No node id found. Make sure to only use a Handle inside a custom Node.

Which seems like this Handle from the custom component is rendering within the "hacky" second provider, even though the whole graph is being rendered correctly in the end, with the edges in the proper place.

Any advise would be appreciated.

### Steps to reproduce the bug or issue

see description above

### Expected behavior

The custom nodes can be configured without issues

**Comment by bcakmakoglu** [comment]
How are you creating the `customNodes` object?
What do your custom nodes look like?
Can you provide more context for this or, better yet, provide a minimal repro of the issue. [You can find a template here.](https://codesandbox.io/p/sandbox/react-flow-starter-ts-r69228)

**Comment by janarvaezkng** [comment]
I'm passing this as a prop to the MyTree component:
```ts
const customNodes = {
  MyNode: {
     NodeComponent: MyNodeComponent,
     size: [100, 200],
     ...
  },
};
```

and then within the library, parsing that object into generating the config like:

```ts
const nodeTypes = {
    MyNode: MyNodeComponent,
}

<ReactFlow nodeTypes={nodeTypes} />
```

This setup works fine if my custom node is created and injected within the library itself, so not sure why when defining the component in a 3rd App and injecting it there fails to work.

I'm suspecting this from the troubleshooting docs is playing a part in this maybe?
> A: You have two different version of @reactflow/core installed.

Could it be that I have to set xyflow/react as a peer dependency instead?

**Comment by sidgawas** [comment]
Any resolution on this issue ? I am getting the same issue when using custom library based on react flow used in 3rd arty app.

**Comment by sidgawas** [comment]
Following work around worked for me. 
From the library I had exported wrapper over ReactFlowProvider as below
`const DesignerProvider: React.FC<DesignerProviderProps> = (props: DesignerProviderProps) => {
    return (<ReactFlowProvider>{props.children}</ReactFlowProvider>);
};

export default DesignerProvider;`

Then in the 3rd party app which used this library I did following thing. 

```
const EditFlowDesigner: React.FC<Props> = () => {
    return (
        <DesignerProvider>
            <ReactFlowProvider>
                <FlowDesigner ></FlowDesigner>
            </ReactFlowProvider>
        </DesignerProvider>
    );
};

Adding ReactFlowProvider again helped me render the flow. But as far as I understand this approach is not the right way. 
```

**Comment by sidgawas** [comment]
Explaining second work-around that sort of worked for me. 
In my case I had defined xyflow/react as peerDependency and then I installed it in 3rd party app. In the custom that was being defined in 3rd party app used useReactFlow() hook from xyflow/react and that sort of created the issue. Somehow it is not able to use DesignerProvider(ReactFlowProvider) defined in library I created. So I defined a wrapper for useReactFlow() callled useFlow and then I used that in my 3rd party library instead of useReactFlow() and that sort of did the job. 

@janarvaezkng In your case you'll have to define a wrapper for Handle  in your library and then import and use it in MyNode.

**Comment by janarvaezkng** [comment]
@sidgawas 🤔 I'll try a few things and update here, although this is still not great. Would be great to have a very small doc about this use case in the react flow site to see if we're doing things correctly.
Thanks for the suggestions, let me see if I can get anywhere less hacky

**Comment by sidgawas** [comment]
Fyi,
This wasn't an issue untill v11 of reactflow. When I migrated to latest version (v12) my library based on reactflow started breaking in 3rd party app.

**Comment by janarvaezkng** [comment]
Yes, same thing happened to me. v11 was fine, v12 started triggering this issue.

I just tried generating a POC build of the library with xyflow/react as a peer dependency, and then installing my library + xyflow in the consumer app, and managed to remove all the hacks with any warnings or errors in the console.

So I guess that's the key: set @xyflow/react as a peer dependency of the library, and then installing everything in the consumer app.

Would still be great to have this documented with an example.

**Comment by moklick** [comment]
Hey @janarvaezkng could you provide a codesandbox for this issue https://new.reactflow.dev?

**Comment by moklick** [comment]
please re-open with a codesandbox or repo.
